### ISSUE TITLE
soup rewind: record which dataset rows were in every training step, so a loss spike can be traced to a named row

### ISSUE THREAD

**Issue by Srinivasan8888** [MakazhanAlpamys/Soup#1000]
When the loss spikes at step 340, Soup can't tell you which dataset row caused it.

- `soup why` reads the tracker's metric series and can report that the loss diverged. Its detector compares the end of the run with the start (`utils/why.py:115`, `_check_explosion`), and nothing in Soup records which rows were in a batch. So it can't name a step, let alone a row.
- `soup adapters blame` attributes weight movement to rows after the fact, using an influence approximation (`utils/blame.py`). That is approximate, gradient-based, and needs the model loaded.
- `loss_watchdog` / `loss_spike_recovery` (#57, #802) roll back and carry on. They don't say what happened.

## Proposal

While an SFT run trains, append one JSON line per micro-batch to `<output>/rewind.jsonl`. Each line holds the dataset row ids in that micro-batch, each row's mean supervised-token loss, and its supervised-token count. Then:

```
soup rewind <run>             # steps whose loss is non-finite or > 2x the median of the previous 20; details the worst
soup rewind <run> --step 340  # rows of that step ranked by share of the step's summed token loss, with a preview
```

The answer comes from what was recorded during training, so it is exact, takes seconds, and needs no model or gradients.

## How, checked against the installed stack (transformers 5.16.1, trl 0.29.1, mlx 0.32.2, mlx-lm 0.31.3)

- **Transformers.** Stock `Trainer` doesn't pass `inputs` to `on_step_end` (`monitoring/callback.py:125` says so), and a row-id column doesn't survive TRL's collator.
  - Row ids come from wrapping the sampler that `_get_train_sampler` returns. That runs in the main process, and I checked it with `dataloader_num_workers=2`.
  - Per-row loss comes from `compute_loss` calling `super()` with `return_outputs=True` and reducing the logits TRL already computed. There is no second forward pass, and TRL's scalar loss is unchanged.
- **MLX.** mlx-lm compiles its training step (`mlx_lm/tuner/trainer.py:248`), so a Python side effect inside the loss runs only while the step is being traced.
  - Per-row losses leave the step through the one channel mlx-lm writes back on every call: arrays reassigned into `optimizer.state`.
  - Row ids come from a `CacheDataset` subclass wrapped around the training dataset. It has to be a subclass: a plain wrapper makes mlx-lm read every row for its length sort, which queued 8 ids before a 2-row batch.
  - MLX ignores `training.seed` (`mlx_sft.py:226`), which is why replaying the sampler was never an option.

## Live check (8 GB M1)

Qwen2.5-0.5B-Instruct-4bit, 64 chat rows, 2 epochs. Row 37's reply was replaced with random base64. `soup rewind`, with no `--step`, printed:

```
warning | step 58 | loss 5.9822 | baseline 0.0005 | 12601.1x
 37 | 6.1954 | 477 tokens | 100.0% | zQcs2L5vn2KsTAnCggbn41WUqms0...
row 37: 100% of step 58's loss, 477 tokens
```

On CPU with `hf-internal-testing/tiny-random-LlamaForCausalLM`, the transformers path records every step. That model's loss is flat (~10.37), so there is no spike to detect, but `--step 6` ranks row 37 first at 83%.

## Scope of the first PR

- `task: sft` on both backends, single process. One new field, `training.rewind_log: bool`, default on. Peak memory was identical with it on and off (3.458 GB on the MLX run above).
- The recorder turns itself off with one warning, and never records wrong rows, in these cases:
  - `resume_from_checkpoint`: accelerate's `SkipBatchSampler` enumerates the wrapped sampler for skipped batches.
  - `packing` / `padding_free`
  - distributed launches
  - Liger's missing logits
- `soup rewind` finds a crashed run's log through its config. The tracker writes `output_dir` only in `finish_run` (`experiment/tracker.py:410`), and a crashed run is exactly when you need the log.
- Replay from a checkpoint (re-run a step's rows, compare the loss then and now) could be a later follow-up.

## Not measured

- Wall-time overhead: single runs on this machine were too noisy to quote a number.
- No CUDA run. I only have an M1.

I'm claiming this. The implementation is written, tested, and checked by breaking each fix to confirm a test fails; the PR will link here.

**Comment by MakazhanAlpamys** [comment]
Closed by #1018 (merged as 0898006c), by @Srinivasan8888. An SFT run now writes `rewind.jsonl` into its output directory. `soup rewind` lists the steps where the loss spiked and ranks the dataset rows behind each one.
